src/history/base.rs

```diff
@@ -14,42 +20,285 @@ pub enum HistoryNavigationQuery {
     // Fuzzy Search
 }
 
-/// Interface of a history datastructure that supports stateful navigation via [`HistoryNavigationQuery`].
+// todo: merge with [HistoryNavigationQuery]
+pub enum CommandLineSearch {
+    Prefix(String),
+    Substring(String),
+    Exact(String),
+}
+
+#[derive(Clone, Copy, PartialEq, Eq)]
+pub enum SearchDirection {
+    Backward,
+    Forward,
+}
+
+pub struct SearchFilter {
+    pub command_line: Option<CommandLineSearch>,
+    pub not_command_line: Option<String>, // to skip duplicates
```

Comment:
I added it because the up-arrow navigation should skip duplicate values, so the search needs to know the previously shown command. I couldn't figure out a better way to do it than just explicitly exclude the currently shown value, because that value does not appear in the returned results themselves so can't really be handled by a `DISTINCT` or `GROUP BY`

Reply:
i changed the comment to make this clearer, since it's not really for deduplication in general

Reply:
Could the `HistoryCursor` do that internallly?